src/container/srv_target.c

```diff
@@ -2104,3 +2104,230 @@ ds_cont_oid_alloc_handler(crt_rpc_t *rpc)
+	return 0;
+}
+
+static void
+cont_ec_xs_reduce_free(struct dss_stream_arg_type *xs)
+{
+	struct cont_ec_xs_query_arg *xs_arg = xs->st_arg;
+
+	D_FREE_PTR(xs_arg->ephs);
+	D_FREE_PTR(xs_arg);
+}
+
+struct cont_ec_eph *
+lookup_insert_cont_ec_eph(struct cont_ec_xs_query_arg *arg, uuid_t cont_uuid)
+{
+	struct cont_ec_eph *found = NULL;
+	int i;
+
+	for (i = 0; i < arg->ephs_cnt; i++) {
+		if (uuid_compare(arg->ephs[i].cont_uuid, cont_uuid) == 0) {
+			found = &arg->ephs[i];
+			return found;
+		}
+	}
+
+	D_ASSERT(arg->ephs_cnt < arg->ephs_alloc_cnt);
+
+	uuid_copy(arg->ephs[arg->ephs_cnt].cont_uuid, cont_uuid);
+	found = &arg->ephs[arg->ephs_cnt];
+	arg->ephs_cnt++;
+	return found;
+}
+
+static void
+cont_ec_eph_reduce(void *agg_arg, void *xs_arg)
+{
+	struct cont_ec_xs_query_arg	*a_arg = agg_arg;
+	struct cont_ec_xs_query_arg	*x_arg = xs_arg;
+	int				i;
+
+	if (a_arg->ephs_alloc_cnt < x_arg->ephs_cnt) {
+		struct cont_ec_eph *new_ephs;
+
+		D_REALLOC_ARRAY(new_ephs, a_arg->ephs, x_arg->ephs_cnt);
+		if (new_ephs == NULL)
+			return;
+		a_arg->ephs_alloc_cnt = x_arg->ephs_cnt;
+		a_arg->ephs = new_ephs;
+	}
+
+	for (i = 0; i < x_arg->ephs_cnt; i++) {
+		struct cont_ec_eph *c_eph;
+
+		c_eph = lookup_insert_cont_ec_eph(a_arg,
+						  x_arg->ephs[i].cont_uuid);
+		if (c_eph->eph == 0 ||
+		    (x_arg->ephs[i].eph != 0 &&
+		     c_eph->eph > x_arg->ephs[i].eph)) {
+			c_eph->eph = x_arg->ephs[i].eph;
+			c_eph->updated = 1;
+		}
+	}
+}
+
+static int
+cont_ec_eph_query_one(void *arg)
+{
+	struct dss_coll_stream_args	*reduce = arg;
+	struct dss_stream_arg_type	*streams = reduce->csa_streams;
+	struct dss_module_info		*info = dss_get_module_info();
+	int				 tid = info->dmi_tgt_id;
+	struct cont_ec_xs_query_arg	*x_arg = streams[tid].st_arg;
+	struct ds_pool_child		*dpc;
+	struct ds_cont_child		*dcc;
+	int				total = 0;
+	struct cont_ec_eph		*ephs;
+	int				i = 0;
+	int				rc = 0;
+
+	dpc = ds_pool_child_lookup(x_arg->pool_uuid);
+	if (dpc == NULL)
+		return -DER_NONEXIST;
+
+	d_list_for_each_entry(dcc, &dpc->spc_cont_list, sc_link)
+		total++;
+
+	if (total == 0)
+		D_GOTO(out, rc = 0);
+
+	D_ALLOC_ARRAY(ephs, total);
+	if (ephs == NULL)
+		D_GOTO(out, rc = -DER_NOMEM);
+
+	d_list_for_each_entry(dcc, &dpc->spc_cont_list, sc_link) {
+		uuid_copy(ephs[i].cont_uuid, dcc->sc_uuid);
+		ephs[i].eph = dcc->sc_ec_agg_eph;
+		i++;
+	}
+
+	x_arg->tgt_id = tid;
+	x_arg->ephs_cnt = total;
+	x_arg->ephs_alloc_cnt = total;
+	x_arg->ephs = ephs;
+
+out:
+	ds_pool_child_put(dpc);
+	return rc;
+}
+
+/**
+ * This ULT is actually per pool to collect all container EC aggregation
+ * epoch, then report to the container service leader.
+ */
+#define EC_TGT_AGG_INTV	 (10ULL * NSEC_PER_SEC)	/* seconds interval to check*/
+void
+ds_cont_tgt_ec_agg_ult(void *data)
+{
+	uuid_t				*pool_uuid = data;
+	struct ds_pool			*pool;
+	struct dss_sleep_ult		*sleep_ult;
+	struct cont_ec_xs_query_arg	query_arg = { 0 };
+	int				rc;
+
+	pool = ds_pool_lookup(*pool_uuid);
```

Comment:
Minor, the life span of this ULT is from pool start to pool stop, and ds_pool is held on pool start, you can use the ds_pool directly without holding it again.

Reply:
oh, ok